stdlib/LinearAlgebra/src/uniformscaling.jl

```diff
@@ -215,17 +215,17 @@ end
 function (+)(A::AbstractMatrix, J::UniformScaling)
     checksquare(A)
     B = copy_oftype(A, Base._return_type(+, Tuple{eltype(A), typeof(J)}))
-    @inbounds for i in axes(A, 1)
-        B[i,i] += J
+    for i in max(first(axes(A, 1)), first(axes(A, 2))):min(last(axes(A, 1)), last(axes(A, 2)))
```

Comment:
Could this just be `for i in intersect(axes(A)...)`?

Reply:
Yes, intersect is exactly the needed function:
https://github.com/JuliaLang/julia/blob/809f27c53df7a54388a687a847e9494e0d29bd4f/base/range.jl#L824

By the way, we can remove checksquare(A) test now. Is a new pull request needed to examine this proposal?

Reply:
For the intersect thing just update the PR. Removing `checksquare` seems like it would merit discussion in a separate PR.

Reply:
Although I guess this PR does raise the question whether this is really the right thing to do. So far, we've basically been looking at UniformScaling as an automatically resizing matrix, so it's not entirely clear what it should do with an OffsetArray. @timholy thoughts?

Reply:
@fp4code could you elaborate on the use case where you wanted this operation? That could perhaps guide intuition as to what it should do.

Reply:
The interest to add a UniformScaling to a rectangular matrix is moderate. Maybe there some risks to offer this kind of freedom. I'm not able to evaluate pros and cons.

Reply:
I think it would be a mistake to remove `checksquare`, and if we're starting to add more OffsetArray support in LInearAlgebra (a good thing in my estimation, see https://github.com/JuliaArrays/OffsetArrays.jl/pull/146), `checksquare` should be modified to check `axes` rather than `size`. The only sensible definition for "the diagonal" is the set `(i, j) where i == j`, not some random "the same number of steps from the left and top of the matrix, whatever they happen to be."

Reply:
There is two choices.
1) extend +I as "fill the diagonal with ones" ; the the constraint "checksquare" can be removed
2) restrict +I to square matrices algebra, and indeed, checksquare needs to be rewritten in order to control the identity of axes

Reply:
> For the intersect thing just update the PR...
Please can you help me about how to update a PR here? With just a new commit, or do you want a rebase?

Reply:
Worth noting that `A = Matrix(I, 3, 4)` does work. But it is a bit more strange for `A + I` to work, because normally you'd expect this to make sense:
```
v = rand(4)
(A + I)*v == A*v + v  # RHS is an error
```
I guess that argues for option 2, making `checksquare(A)` check the axes. In which case there is no need to alter `for i in axes(A, 1)`.

Reply:
> Please can you help me about how to update a PR here? With just a new commit, or do you want a rebase?

It doesn't really matter since the PR is likely to get squashed upon merge. Might be easiest to just add a commit.

Reply:
> Is `checksquare` used in other contexts where allowing different index ranges on left & right would make sense?

Not really. But `checksquare` is problematic now, because it is used to return the size and not the common axis range. Worst, the multiple argument `checksquare`, used in lapack.jl, is not efficient. See here for example https://github.com/JuliaLang/julia/blob/3074306fb12746931c9ba3774a41c495e7be2779/stdlib/LinearAlgebra/src/lapack.jl#L6152 
The multiple arguments `checksquare` https://github.com/JuliaLang/julia/blob/3074306fb12746931c9ba3774a41c495e7be2779/stdlib/LinearAlgebra/src/LinearAlgebra.jl#L227 is called, creating a vector by several inefficients `push!` calls. After this call, lapack function is checking that all matrices have the same dimensions. Well, the `checksquare(A...)` need to be replaced in lapack.jl by a more efficient function, `get_common_size_of_square_matrices(A...)`. And here in uniformscaling `checksquare(A)` could be replaced by `r = get_range_of_square_matrix(A)`, replacing `for i in axes(A, 1)` by `for i in r`.